docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,358 @@
+endpoints, where credentials should be obtained from an `.htpasswd` file, using
+the same format as [Apache `.htpasswd` files][htpasswd], whose location is set
+in the Tendermint configuration file.
+
+`.htpasswd` files are relatively standard and are supported by many web servers.
+The format is relatively straightforward to parse and interpret.
+
+We should strongly consider only supporting the bcrypt encryption option for
+passwords stored in `.htpasswd` files.
+
+### Configuration
+
+The following configuration file update is proposed to support the data
+companion API.
+
+```toml
+# A data companion, if enabled, is intended to offload the storage of certain
+# types of data from the node. Specifically:
+# 1. Block data, including transactions.
+# 2. Block results, including events, transaction results, validator and
+#    consensus parameter updates.
+#
+# A data companion can influence the pruning height of the node, and therefore
+# the data companion gRPC service is considered to be a sensitive endpoint that
+# is password-protected by default.
+[data_companion]
+
+# Is the data companion gRPC API enabled at all? Default: false
+enabled = false
+
+# The maximum number of blocks to allow the data companion's retain height to
+# lag behind the current height. The node will block all other operations until
+# the companion's retain height is less than `max_retain_lag` blocks behind the
+# current height. Default: 100
+max_retain_lag = 100
+
+# Authentication configuration for the data companion.
+[data_companion.authentication]
+
+# The authentication method to use. At present, the only supported method is
+# "basic" (i.e. basic HTTP authentication).
+method = "basic"
+
+# Path to the file containing basic authentication credentials to access the
+# data companion service. If the data companion is enabled and this password
+# file is not supplied, or the file does not exist, or the file is of an invalid
+# format, the node will fail to start.
+#
+# See https://httpd.apache.org/docs/current/programs/htpasswd.html for details
+# on how to create/configure this file.
+password_file = "/path/to/.htpasswd"
+```
+
+### Metrics
+
+The following metrics are proposed to be added to monitor the health of the
+interaction between a node and its data companion:
+
+- `data_companion_retain_height` - The current retain height as requested by the
+  data companion. This can give operators insight into whether the companion is
+  lagging significantly behind the current network height.
+
+## Consequences
+
+### Positive
+
+- Facilitates offloading of data to an external service, which can be scaled
+  independently of the node
+  - Potentially reduces load on the node itself
+  - Paves the way for eventually reducing the surface area of a node's exposed
+    APIs
+- Allows the data companion more leeway in reading the data it needs than the
+  approach in [ADR 100][adr-100]
+- Simpler implementation than [ADR 100][adr-100]
```

Comment:
It's hard to give a very accurate estimate here without actually implementing both, but ADR-101 is probably going to be far simpler in its implementation because the majority of the functionality is effectively provided by #81.

With ADR-100, it still needs to explicitly cater for crash recovery (sending the last block upon startup), and the gRPC implementation is probably going to be more complex than ADR-101.